File: backend/routers/sse.py
"""
SSE (Server-Sent Events) endpoint for real-time streaming.

Provides:
- POST /rooms/{room_id}/stream/ticket - Get a single-use SSE auth ticket
- GET /rooms/{room_id}/stream?ticket=... - SSE event stream
"""

import asyncio
import json
import logging

from fastapi import APIRouter, HTTPException, Request, status
from infrastructure.sse import EventBroadcaster
from infrastructure.sse_ticket import SSETicketManager
from sse_starlette.sse import EventSourceResponse

logger = logging.getLogger(__name__)

router = APIRouter()

KEEPALIVE_INTERVAL = 15  # seconds


def _get_broadcaster(request: Request) -> EventBroadcaster:
    return request.app.state.event_broadcaster


def _get_ticket_manager(request: Request) -> SSETicketManager:
    return request.app.state.sse_ticket_manager


@router.post("/{room_id}/stream/ticket")
async def create_stream_ticket(room_id: int, request: Request):
    """
    Create a single-use ticket for SSE authentication.

    Requires JWT auth (X-API-Key header). The returned ticket is used
    as a query parameter for the SSE endpoint since EventSource API
    cannot send custom headers.
    """
    # Auth info is injected by middleware into request.state
    user_id = getattr(request.state, "user_id", None)
    user_role = getattr(request.state, "user_role", None)

    if not user_id:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Not authenticated")

    ticket_manager = _get_ticket_manager(request)
    ticket = ticket_manager.create_ticket(user_id=user_id, role=user_role or "admin", room_id=room_id)

    return {"ticket": ticket}


@router.get("/{room_id}/stream")
async def stream_events(room_id: int, ticket: str, request: Request):
    """
    SSE endpoint for real-time event streaming.

    Authenticated via single-use ticket (from POST /stream/ticket).
    Streams events: stream_start, content_delta, thinking_delta, stream_end, new_message, keepalive.
    """
    ticket_manager = _get_ticket_manager(request)
    ticket_data = ticket_manager.validate_ticket(ticket, room_id)

    if not ticket_data:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid or expired ticket")

    broadcaster = _get_broadcaster(request)

    # Get streaming state manager for catch-up events
    agent_manager = request.app.state.agent_manager
    streaming_state = await agent_manager.get_streaming_state_for_room(room_id)

    async def event_generator():
        queue = broadcaster.subscribe(room_id)
        try:
            # Send catch-up events for agents already streaming
            if streaming_state:
                for agent_id, state in streaming_state.items():
                    catch_up = {
                        "type": "catch_up",
                        "agent_id": agent_id,
                        "agent_name": state.get("agent_name", ""),
                        "thinking_text": state.get("thinking_text", ""),
                        "response_text": state.get("response_text", ""),
                    }
                    yield {"event": "catch_up", "data": json.dumps(catch_up)}

            # Send connected confirmation
            yield {"event": "connected", "data": json.dumps({"room_id": room_id})}

            while not broadcaster.shutdown_event.is_set():
                try:
                    data = await asyncio.wait_for(queue.get(), timeout=KEEPALIVE_INTERVAL)
                    if data is None:
                        # Shutdown sentinel
                        break
                    # data is already a JSON string from broadcaster
                    parsed = json.loads(data)
                    event_type = parsed.get("type", "message")
                    yield {"event": event_type, "data": data}
                except asyncio.TimeoutError:
                    # Send keepalive to prevent connection timeout
                    yield {"event": "keepalive", "data": ""}
                except asyncio.CancelledError:
                    break
        finally:
            broadcaster.unsubscribe(room_id, queue)

    return EventSourceResponse(event_generator())

File: backend-ts/src/http/routes/rooms/sse.ts
/**
 * SSE streaming — port of `backend/routers/sse.py`.
 *
 * Two routes that only make sense together:
 *
 *   POST /rooms/:room_id/stream/ticket   JWT in, single-use ticket out
 *   GET  /rooms/:room_id/stream?ticket=  the stream, authenticated by that ticket
 *
 * The split exists because `EventSource` cannot set a request header, so the
 * long-lived JWT would otherwise have to travel in the query string and land in
 * access logs. `middleware/auth.ts` excludes the GET for that reason and this
 * module authenticates it instead — the exclusion and the `validateTicket` call
 * below are two halves of one check, and removing either opens the stream.
 *
 * ## Gap against Python
 *
 * Python replays *catch-up* events on connect, so a client that reloads mid-turn
 * immediately sees the partial text an agent has already produced. It reads them
 * from `AgentManager.get_streaming_state_for_room`, a per-room registry of
 * in-flight response text. The TypeScript SDK keeps that state on the turn
 * rather than on a manager and nothing publishes it per room yet — the same gap
 * `routes/game/polling.ts` documents for `thinking_text`. The stream is correct
 * from the moment it connects; it just does not backfill.
 */

import { Hono } from 'hono'

import { KEEPALIVE_INTERVAL_MS } from '../../../infrastructure/sse'
import { getLogger } from '../../../infrastructure/logging/logger'
import { HttpError } from '../../errors'
import { identityOf, type AppState } from '../../state'
import type { AppEnv } from '../../types'
import { intPathParam } from '../game/shared'
import { ensureRoomAccessFor } from './shared'

const logger = getLogger('SSERouter')

/**
 * Render one SSE frame.
 *
 * `data` is written as a single line because every payload here is JSON, which
 * never contains a raw newline. A multi-line payload would need one `data:` per
 * line, and silently truncating at the first newline is the classic SSE bug.
 */
function frame(event: string, data: string): string {
  return `event: ${event}\ndata: ${data}\n\n`
}

export function createSseRoutes(state: AppState): Hono<AppEnv> {
  const routes = new Hono<AppEnv>()

  /**
   * Trade a JWT for a ticket.
   *
   * Reached through the normal auth middleware, so `identityOf` is already the
   * verified caller. The room access check is *not* in Python — its handler
   * only reads `request.state` — but the ticket it mints is a bearer credential
   * for a room's entire event stream, and issuing one for a room the caller
   * cannot otherwise read would be a way around `ensure_room_access`.
   */
  routes.post('/rooms/:room_id/stream/ticket', (c) => {
    const roomId = intPathParam(c, 'room_id')
    ensureRoomAccessFor(c, state.db, roomId)

    const identity = identityOf(c)
    const ticket = state.tickets.createTicket(identity.userId, identity.role, roomId)
    return c.json({ ticket })
  })

  /**
   * The event stream.
   *
   * Runs until the client goes away, the server shuts down, or the room's
   * broadcaster ends the queue. `ReadableStream.from` drives the generator, so
   * a disconnect runs its `finally` — which is what unsubscribes; there is no
   * separate disconnect handler to keep in step with it.
   */
  routes.get('/rooms/:room_id/stream', (c) => {
    const roomId = intPathParam(c, 'room_id')

    const ticket = c.req.query('ticket')
    const ticketData = ticket ? state.tickets.validateTicket(ticket, roomId) : null
    // One message for "absent", "expired", "already spent" and "wrong room":
    // distinguishing them would tell an attacker which half to vary.
    if (ticketData === null) throw new HttpError(401, 'Invalid or expired ticket')

    const queue = state.broadcaster.subscribe(roomId)
    logger.debug(`SSE stream opened for room ${roomId} by ${ticketData.userId}`)

    async function* events(): AsyncGenerator<string> {
      try {
        yield frame('connected', JSON.stringify({ room_id: roomId }))

        while (!state.broadcaster.isShuttingDown) {
          let data: string | null
          try {
            // The timeout is the keepalive clock: parking forever would let an
            // idle connection be reaped by any proxy in the path.
            data = await queue.next(AbortSignal.timeout(KEEPALIVE_INTERVAL_MS))
          } catch {
            // Only the timeout can land here — the queue itself resolves with
            // null rather than throwing when it ends.
            yield frame('keepalive', '')
            continue
          }

          // Null means the queue ended: unsubscribed, or server shutdown.
          if (data === null) break

          // The broadcaster serialises the event; `type` doubles as the SSE
          // event name so the frontend's per-name listeners fire.
          let eventName = 'message'
          try {
            const parsed = JSON.parse(data) as { type?: unknown }
            if (typeof parsed.type === 'string') eventName = parsed.type
          } catch {
            // Unreachable via `broadcast`, which serialises what it sends.
          }
          yield frame(eventName, data)
        }
      } finally {
        state.broadcaster.unsubscribe(roomId, queue)
        logger.debug(`SSE stream closed for room ${roomId}`)
      }
    }

    // Driven by an explicit pull rather than `ReadableStream.from`, which Bun
    // implements but these typings do not declare. `cancel` is what makes a
    // client disconnect run the generator's `finally`, and that is what
    // unsubscribes — so there is no separate disconnect path to keep in step.
    const iterator = events()[Symbol.asyncIterator]()
    const stream = new ReadableStream<string>({
      async pull(controller) {
        const { value, done } = await iterator.next()
        if (done) {
          controller.close()
          return
        }
        controller.enqueue(value)
      },
      async cancel(reason) {
        await iterator.return?.(reason)
      },
    })

    return new Response(stream.pipeThrough(new TextEncoderStream()), {
      headers: {
        'content-type': 'text/event-stream; charset=utf-8',
        'cache-control': 'no-store',
        // Nginx buffers proxied responses by default, which would hold every
        // delta until the response ended — i.e. forever. Python's sse_starlette
        // sends this for the same reason.
        'x-accel-buffering': 'no',
      },
    })
  })

  return routes
}
